dio/lib/src/transformers/sync_transformer.dart

```diff
@@ -42,86 +42,96 @@ class SyncTransformer extends Transformer {
     }
   }
 
-  /// As an agreement, we return the [response] when the
-  /// Options.responseType is [ResponseType.stream].
   @override
   Future<dynamic> transformResponse(
     RequestOptions options,
-    ResponseBody response,
+    ResponseBody responseBody,
   ) async {
+    final responseType = options.responseType;
+    // Do not handled the body for streams.
     if (options.responseType == ResponseType.stream) {
-      return response;
+      return responseBody;
     }
-    int length = 0;
-    int received = 0;
+
     final showDownloadProgress = options.onReceiveProgress != null;
+    final int totalLength;
     if (showDownloadProgress) {
-      length = int.parse(
-        response.headers[Headers.contentLengthHeader]?.first ?? '-1',
+      totalLength = int.parse(
+        responseBody.headers[Headers.contentLengthHeader]?.first ?? '-1',
       );
+    } else {
+      totalLength = 0;
     }
-    final completer = Completer();
-    final stream = response.stream.transform<Uint8List>(
+
+    int received = 0;
+    final stream = responseBody.stream.transform<Uint8List>(
       StreamTransformer.fromHandlers(
         handleData: (data, sink) {
           sink.add(data);
           if (showDownloadProgress) {
             received += data.length;
-            options.onReceiveProgress?.call(received, length);
+            options.onReceiveProgress?.call(received, totalLength);
           }
         },
       ),
     );
+
+    final streamCompleter = Completer<void>();
+    int finalLength = 0;
     // Keep references to the data chunks and concatenate them later.
     final chunks = <Uint8List>[];
-    int finalSize = 0;
-    final StreamSubscription subscription = stream.listen(
+    final subscription = stream.listen(
       (chunk) {
-        finalSize += chunk.length;
+        finalLength += chunk.length;
         chunks.add(chunk);
       },
       onError: (Object error, StackTrace stackTrace) {
-        completer.completeError(error, stackTrace);
+        streamCompleter.completeError(error, stackTrace);
+      },
+      onDone: () {
+        streamCompleter.complete();
       },
-      onDone: () => completer.complete(),
       cancelOnError: true,
     );
     options.cancelToken?.whenCancel.then((_) {
       return subscription.cancel();
```

Comment:
I wasn't sure about how this line will behave (whether we should complete the completer or not). It's a void future and we await for the completer in the outside scope.